noxfile.py

```diff
@@ -0,0 +1,285 @@
+        "stubs/",
+    )
+
+
+def _repl_version(session: nox.Session, new_version: str):
+    with open("pyproject.toml", "r+") as fp:
+        lines = tuple(fp)
+        fp.seek(0)
+        for line in lines:
+            if line.startswith("version = "):
+                line = f'version = "{new_version}"\n'
+            fp.write(line)
+
+
+def check_no_modifications(session: nox.Session) -> None:
+    modified = session.run(
+        "git",
+        "status",
+        "--porcelain=v1",
+        "--untracked=normal",
+        external=True,
+        silent=True,
+    )
+    if modified:
+        session.error(
+            "There are modified or untracked files. "
+            "Commit, restore, or remove them before running this"
+        )
+
+
+@nox.session
+def bump(session: nox.Session):
+    check_no_modifications(session)
+    if len(session.posargs) not in (1, 2):
+        session.error(
+            "Must specify 1-2 positional arguments: nox -e bump -- <version> "
+            "[ <release_summary_message> ]."
+            " If release_summary_message has not been specified, "
+            "a file changelogs/fragments/<version>.yml must exist"
+        )
+    version = session.posargs[0]
+    fragment_file = Path(f"changelogs/fragments/{version}.yml")
+    if len(session.posargs) == 1:
+        if not fragment_file.is_file():
+            session.error(
+                f"Either {fragment_file} must already exist, "
+                "or two positional arguments must be provided."
+            )
+    install(session, "antsibull-changelog", "tomli ; python_version < '3.11'")
+    _repl_version(session, version)
+    if len(session.posargs) > 1:
+        fragment = session.run(
+            "python",
+            "-c",
+            "import yaml ; "
+            f"print(yaml.dump(dict(release_summary={repr(session.posargs[1])})))",
+            silent=True,
+        )
+        with open(fragment_file, "w") as fp:
+            print(fragment, file=fp)
+        session.run("git", "add", "pyproject.toml", fragment_file, external=True)
+        session.run("git", "commit", "-m", f"Prepare {version}.", external=True)
+    session.run("antsibull-changelog", "release")
+    session.run(
+        "git",
+        "add",
+        "CHANGELOG.rst",
+        "changelogs/changelog.yaml",
+        "changelogs/fragments/",
+        external=True,
+    )
+    install(session, ".")  # Smoke test
+    session.run("git", "commit", "-m", f"Release {version}.", external=True)
+    session.run(
+        "git",
+        "tag",
+        "-a",
+        "-m",
+        f"antsibull {version}",
+        "--edit",
+        version,
+        external=True,
+    )
+
+
+@nox.session
+def publish(session: nox.Session):
+    check_no_modifications(session)
+    install(session, "hatch")
+    session.run("hatch", "publish", *session.posargs)
+    session.run("git", "push", "--follow-tags")
+    version = session.run("hatch", "version", silent=True).strip()
```

Comment:
Consider using https://github.com/ofek/hatch-vcs

Reply:
I'm weird in that I somewhat like hardcoding the version. I prefer to avoid dependencies on a VCS system that are not actually needed. If I `rm -rf .git`, `pip install .` should still work. Also, I don't like that hatch-vcs pulls in setuptools_scm which pulls in setuptools[^1]. However, if others are really on board with hatch-vcs, I'd consider it.

[^1]: I remember reading somewhere that Ronny was considering splitting out setuptools_scm's backend in a separate package at some point.

Reply:
I also prefer having the version hardcoded. (As long as it's hardcoded in only one place :) .)

Reply:
> 1. I remember reading somewhere that Ronny was considering splitting out setuptools_scm's backend in a separate package at some point.

That would be interesting. Though, there's https://github.com/mtkennerly/dunamai for that. I probably need to point Ronny at it...

Reply:
> (With one place, I mean one place inside the file tree of a commit of the repo. And not in the repo tag/branch names.)

@felixfontein note that setutools-scm allows writing the version on disk, into a Python file so the source code can still just import it, or read from the dist metadata.

@gotmax23 I don't think there's a reasonable use-case for `rm -rf .git`. And installs from sdist and Git archives come with that version placed in the right file anyway.

Reply:
> `pip install .` should still work

It would work, just produce a dummy version.